### DISCUSSION TITLE
Migrating to CSLA 8, now ObjectDisposedException for IServiceProvider

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#4135] [Questions]
I'm in the process of migrating a Blazor project from .NET 7.0 + CSLA 7.0.4 to .NET 8.0 + CSLA 8.2.5.

After a bunch of DAL interfaces and factories are registered with `builder.Services.AddTransient` some lists are loaded in a `Parallel.ForEach` before `app.Run()`.

This has been working for a long time with .NET 7.0, but now it's throwing `System.ObjectDisposedException` and `IServiceProvider` is the ObjectName. Disabling the parallel for with `maxParallelism = 1` doesn't fix it.

I've read [Upgrading to CSLA 8](https://github.com/MarimerLLC/csla/blob/main/docs/Upgrading%20to%20CSLA%208.md) and looked at `Program.cs` in the BlazorExample to see if I'm missing something. Can anyone point me in the right direction?

**Comment by rockfordlhotka** [comment]
This is before app.run but after the DI root provider has been created?

Are you using a local or remote data portal?

**Reply by michaelcsikos** [reply]
Yes, immediately before `app.Run()`.

Blazor Server with server-side data portal:
```cs
var builder = WebApplication.CreateBuilder(args);

builder.Logging.AddConfiguration(builder.Configuration.GetSection("Logging"));

builder.Services.AddRazorPages();
builder.Services.AddServerSideBlazor();
builder.Services.AddCascadingAuthenticationState();
builder.Services.AddHttpContextAccessor();

builder.Services.AddAuthentication(options =>
{
    options.DefaultSignInScheme       = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
    options.DefaultChallengeScheme    = CookieAuthenticationDefaults.AuthenticationScheme;
}).AddCookie();

builder.Services.AddCsla(o => o.AddAspNetCore()
                               .AddServerSideBlazor(o => o.UseInMemoryApplicationContextManager = false)
                               .Security           (s => s.FlowSecurityPrincipalFromClient      = true)
                               .DataPortal         (d => d.AddServerSideDataPortal()));
```

It seems you can't run anything after `app.Run()`.

**Reply by michaelcsikos** [reply]
I'll admit I do get a bit confused with all the code in `Program.cs`. Do I need Razor components or pages, interactive server components, render mode, razor hub, controllers, etc, etc...

**Comment by ossendorf-at-hoelscher** [comment]
That won't work as the exception indicates. You must b using a ServiceProvider which gets disposed before the actual application runs. And since _all_ business objects get the service provider from the DI scope they are in they are _not_ in the scope of your app. So disposing the scope to retrieve the lists effectivly kills all retrieved lists due to the now disposed srvice provider.
You have to retrieve the lists with the apps service provider.

**Comment by michaelcsikos** [comment]
The problem was the `Parallel.ForEach` which doesn't `await` the tasks. Using `Task.WhenAll(tasks)` works, and I assume `Parallel.ForEachAsync` would work, too. I'm just not sure why it wasn't throwing in .NET 7.
